src/service/service.ts

```diff
@@ -300,18 +300,28 @@ export class Discv5 extends (EventEmitter as { new (): Discv5EventEmitter }) {
   }
 
   /**
-   * Broadcast TALKREQ message to all nodes in routing table
+   * Broadcast TALKREQ message to all nodes in routing table and returns response
    */
-  public async broadcastTalkReq(payload: Buffer, protocol: string | Uint8Array): Promise<void> {
-    const msg = createTalkRequestMessage(payload, protocol);
-    for (const node of this.kadValues()) {
-      const sendStatus = this.sendRequest(node.nodeId, msg);
-      if (!sendStatus) {
-        log(`Failed to send TALKREQ message to node ${node.nodeId}`);
-      } else {
-        log(`Sent TALKREQ message to node ${node.nodeId}`);
+  public async broadcastTalkReq(payload: Buffer, protocol: string | Uint8Array, timeout = 1000): Promise<Buffer> {
+    return await new Promise((resolve, reject) => {
+      const msg = createTalkRequestMessage(payload, protocol);
+      const responseTimeout = setTimeout(() => reject("Request timed out"), timeout);
+      this.on("talkRespReceived", (srcId, enr, res) => {
```

Comment:
We need to remove the event listener in either case so they don't build up.

Reply:
Does it make sense that a node would ever have more than one `TALKREQ` message waiting for a response?  It's not 100% obvious to me that we would ever want to send more than one at a time but if so, that makes clearing the right listener a little trickier.  I was contemplating creating a dynamic message type of `TALKREQ-${messageId}`in this instance as the most straightforward way to track `TALKREQ` messages using the promise based interface vs just broadcasting messages out and listening for general responses.  That would change the API somewhat though.  In this case, I'm thinking to revert my changes on this message and create a new `broadcastTalkReqMessagePromise` or something similar that uses the new dynamically named message type so that I can clear the listeners when the promise resolves/rejects.  Thoughts?